File: archive/pulumi-py/Pr7062/lib/monitoring_component.py
"""
CloudWatch Monitoring Component.
Creates CloudWatch alarms with environment-specific thresholds.
"""

from typing import Optional, Dict
import pulumi
import pulumi_aws as aws
from pulumi import ResourceOptions, Output
from lib.environment_config import EnvironmentConfig


class MonitoringComponentArgs:
    """Arguments for Monitoring Component."""

    def __init__(
        self,
        environment_suffix: str,
        env_config: EnvironmentConfig,
        lambda_function_name: Output[str],
        dynamodb_table_name: Output[str],
        tags: Optional[Dict[str, str]] = None
    ):
        self.environment_suffix = environment_suffix
        self.env_config = env_config
        self.lambda_function_name = lambda_function_name
        self.dynamodb_table_name = dynamodb_table_name
        self.tags = tags or {}


class MonitoringComponent(pulumi.ComponentResource):
    """
    Reusable monitoring component with CloudWatch alarms.
    Creates alarms with environment-specific thresholds.
    """

    def __init__(
        self,
        name: str,
        args: MonitoringComponentArgs,
        opts: Optional[ResourceOptions] = None
    ):
        super().__init__('payment:monitoring:MonitoringComponent', name, None, opts)

        child_opts = ResourceOptions(parent=self)

        # Lambda error alarm
        self.lambda_error_alarm = aws.cloudwatch.MetricAlarm(
            f"payment-lambda-errors-{args.environment_suffix}",
            name=f"payment-lambda-errors-{args.environment_suffix}",
            comparison_operator="GreaterThanThreshold",
            evaluation_periods=1,
            metric_name="Errors",
            namespace="AWS/Lambda",
            period=300,  # 5 minutes
            statistic="Sum",
            threshold=args.env_config.lambda_error_alarm_threshold,
            alarm_description=(
                f"Lambda errors exceeded "
                f"{args.env_config.lambda_error_alarm_threshold} "
                f"in {args.env_config.name}"
            ),
            dimensions={
                'FunctionName': args.lambda_function_name
            },
            tags={
                **args.tags,
                'Name': f"payment-lambda-errors-{args.environment_suffix}",
            },
            opts=child_opts
        )

        # DynamoDB read throttle alarm
        self.dynamodb_read_throttle_alarm = aws.cloudwatch.MetricAlarm(
            f"payment-dynamodb-read-throttle-{args.environment_suffix}",
            name=f"payment-dynamodb-read-throttle-{args.environment_suffix}",
            comparison_operator="GreaterThanThreshold",
            evaluation_periods=1,
            metric_name="ReadThrottleEvents",
            namespace="AWS/DynamoDB",
            period=300,  # 5 minutes
            statistic="Sum",
            threshold=args.env_config.dynamodb_throttle_alarm_threshold,
            alarm_description=(
                f"DynamoDB read throttle exceeded "
                f"{args.env_config.dynamodb_throttle_alarm_threshold} "
                f"in {args.env_config.name}"
            ),
            dimensions={
                'TableName': args.dynamodb_table_name
            },
            tags={
                **args.tags,
                'Name': f"payment-dynamodb-read-throttle-{args.environment_suffix}",
            },
            opts=child_opts
        )

        # DynamoDB write throttle alarm
        self.dynamodb_write_throttle_alarm = aws.cloudwatch.MetricAlarm(
            f"payment-dynamodb-write-throttle-{args.environment_suffix}",
            name=f"payment-dynamodb-write-throttle-{args.environment_suffix}",
            comparison_operator="GreaterThanThreshold",
            evaluation_periods=1,
            metric_name="WriteThrottleEvents",
            namespace="AWS/DynamoDB",
            period=300,  # 5 minutes
            statistic="Sum",
            threshold=args.env_config.dynamodb_throttle_alarm_threshold,
            alarm_description=(
                f"DynamoDB write throttle exceeded "
                f"{args.env_config.dynamodb_throttle_alarm_threshold} "
                f"in {args.env_config.name}"
            ),
            dimensions={
                'TableName': args.dynamodb_table_name
            },
            tags={
                **args.tags,
                'Name': f"payment-dynamodb-write-throttle-{args.environment_suffix}",
            },
            opts=child_opts
        )

        # Register outputs
        self.register_outputs({
            'lambda_error_alarm_arn': self.lambda_error_alarm.arn,
            'dynamodb_read_throttle_alarm_arn': self.dynamodb_read_throttle_alarm.arn,
            'dynamodb_write_throttle_alarm_arn': self.dynamodb_write_throttle_alarm.arn,
        })

File: archive/pulumi-ts/Pr6618/lib/components/monitoring-component.ts
import * as aws from '@pulumi/aws';
import * as pulumi from '@pulumi/pulumi';

export interface MonitoringComponentArgs {
  environmentSuffix: string;
  clusterName: pulumi.Input<string>;
  serviceName: pulumi.Input<string>;
  albArn: pulumi.Input<string>;
  targetGroupArn: pulumi.Input<string>;
  rdsClusterIdentifier: pulumi.Input<string>;
}

export class MonitoringComponent extends pulumi.ComponentResource {
  public readonly dashboard: aws.cloudwatch.Dashboard;
  public readonly snsTopic: aws.sns.Topic;
  public readonly alarms: aws.cloudwatch.MetricAlarm[];

  constructor(
    name: string,
    args: MonitoringComponentArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('custom:monitoring:MonitoringComponent', name, {}, opts);

    // Create SNS Topic for Alerts
    this.snsTopic = new aws.sns.Topic(
      `alerts-topic-${args.environmentSuffix}`,
      {
        name: `alerts-topic-${args.environmentSuffix}-pw`,
        displayName: `Alerts for ${args.environmentSuffix} environment`,
        tags: {
          Name: `alerts-topic-${args.environmentSuffix}`,
          Environment: args.environmentSuffix,
        },
      },
      { parent: this }
    );

    // Create CloudWatch Dashboard
    this.dashboard = new aws.cloudwatch.Dashboard(
      `dashboard-${args.environmentSuffix}`,
      {
        dashboardName: `trading-platform-${args.environmentSuffix}-pw`,
        dashboardBody: pulumi
          .all([
            args.clusterName,
            args.serviceName,
            args.albArn,
            args.targetGroupArn,
            args.rdsClusterIdentifier,
          ])
          .apply(([cluster, service, albArn, _targetGroup, rdsCluster]) =>
            JSON.stringify({
              widgets: [
                {
                  type: 'metric',
                  properties: {
                    metrics: [
                      [
                        'AWS/ECS',
                        'CPUUtilization',
                        'ClusterName',
                        cluster,
                        'ServiceName',
                        service,
                      ],
                      ['.', 'MemoryUtilization', '.', '.', '.', '.'],
                    ],
                    period: 300,
                    stat: 'Average',
                    region: aws.config.region,
                    title: 'ECS Service Metrics',
                  },
                },
                {
                  type: 'metric',
                  properties: {
                    metrics: [
                      [
                        'AWS/ApplicationELB',
                        'TargetResponseTime',
                        'LoadBalancer',
                        albArn.split(':').pop(),
                      ],
                      ['.', 'RequestCount', '.', '.'],
                      ['.', 'HTTPCode_Target_2XX_Count', '.', '.'],
                      ['.', 'HTTPCode_Target_5XX_Count', '.', '.'],
                    ],
                    period: 300,
                    stat: 'Sum',
                    region: aws.config.region,
                    title: 'ALB Metrics',
                  },
                },
                {
                  type: 'metric',
                  properties: {
                    metrics: [
                      [
                        'AWS/RDS',
                        'CPUUtilization',
                        'DBClusterIdentifier',
                        rdsCluster,
                      ],
                      ['.', 'DatabaseConnections', '.', '.'],
                      ['.', 'FreeableMemory', '.', '.'],
                    ],
                    period: 300,
                    stat: 'Average',
                    region: aws.config.region,
                    title: 'RDS Aurora Metrics',
                  },
                },
              ],
            })
          ),
      },
      { parent: this }
    );

    // Create CloudWatch Alarms
    this.alarms = [];

    // ECS CPU Alarm
    const ecsCpuAlarm = new aws.cloudwatch.MetricAlarm(
      `ecs-cpu-alarm-${args.environmentSuffix}`,
      {
        name: `ecs-cpu-high-${args.environmentSuffix}`,
        comparisonOperator: 'GreaterThanThreshold',
        evaluationPeriods: 2,
        metricName: 'CPUUtilization',
        namespace: 'AWS/ECS',
        period: 300,
        statistic: 'Average',
        threshold: 80,
        alarmDescription: 'ECS CPU utilization is too high',
        alarmActions: [this.snsTopic.arn],
        dimensions: {
          ClusterName: args.clusterName,
          ServiceName: args.serviceName,
        },
        tags: {
          Name: `ecs-cpu-alarm-${args.environmentSuffix}`,
          Environment: args.environmentSuffix,
        },
      },
      { parent: this }
    );
    this.alarms.push(ecsCpuAlarm);

    // RDS CPU Alarm
    const rdsCpuAlarm = new aws.cloudwatch.MetricAlarm(
      `rds-cpu-alarm-${args.environmentSuffix}`,
      {
        name: `rds-cpu-high-${args.environmentSuffix}`,
        comparisonOperator: 'GreaterThanThreshold',
        evaluationPeriods: 2,
        metricName: 'CPUUtilization',
        namespace: 'AWS/RDS',
        period: 300,
        statistic: 'Average',
        threshold: 80,
        alarmDescription: 'RDS CPU utilization is too high',
        alarmActions: [this.snsTopic.arn],
        dimensions: {
          DBClusterIdentifier: args.rdsClusterIdentifier,
        },
        tags: {
          Name: `rds-cpu-alarm-${args.environmentSuffix}`,
          Environment: args.environmentSuffix,
        },
      },
      { parent: this }
    );
    this.alarms.push(rdsCpuAlarm);

    // ALB 5XX Alarm
    const alb5xxAlarm = new aws.cloudwatch.MetricAlarm(
      `alb-5xx-alarm-${args.environmentSuffix}`,
      {
        name: `alb-5xx-errors-${args.environmentSuffix}`,
        comparisonOperator: 'GreaterThanThreshold',
        evaluationPeriods: 1,
        metricName: 'HTTPCode_Target_5XX_Count',
        namespace: 'AWS/ApplicationELB',
        period: 300,
        statistic: 'Sum',
        threshold: 10,
        alarmDescription: 'ALB is returning too many 5XX errors',
        alarmActions: [this.snsTopic.arn],
        dimensions: {
          LoadBalancer: args.albArn,
        },
        tags: {
          Name: `alb-5xx-alarm-${args.environmentSuffix}`,
          Environment: args.environmentSuffix,
        },
      },
      { parent: this }
    );
    this.alarms.push(alb5xxAlarm);

    this.registerOutputs({
      dashboardName: this.dashboard.dashboardName,
      snsTopicArn: this.snsTopic.arn,
    });
  }
}
